internal/core/src/futures/future_c.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include <memory>
+
+#include "future_c.h"
+#include "folly/init/Init.h"
+#include "Future.h"
+
+extern "C" CFollyInit*
+folly_init() {
+    auto opts = folly::InitOptions{};
+    opts.useGFlags(false);
+    auto initializer = new folly::Init(nullptr, nullptr, opts);
+    return static_cast<CFollyInit*>(static_cast<void*>(initializer));
+}
+
+extern "C" void
+folly_init_destroy(CFollyInit* init) {
+    delete static_cast<folly::Init*>(static_cast<void*>(init));
+}
+
+extern "C" void
+future_cancel(CFuture* future) {
+    static_cast<milvus::futures::IFuture*>(static_cast<void*>(future))
+        ->cancel();
+}
+
+extern "C" bool
+future_is_ready(CFuture* future) {
+    return static_cast<milvus::futures::IFuture*>(static_cast<void*>(future))
+        ->isReady();
+}
+
+extern "C" void
+future_register_ready_callback(CFuture* future,
+                               CUnlockGoMutexFn unlockFn,
+                               CLockedGoMutex* mutex) {
+    static_cast<milvus::futures::IFuture*>(static_cast<void*>(future))
+        ->registerReadyCallback(unlockFn, mutex);
+}
+
+extern "C" void
+future_register_releasable_callback(CFuture* future,
+                                    CUnlockGoMutexFn unlockFn,
+                                    CLockedGoMutex* mutex) {
+    static_cast<milvus::futures::IFuture*>(static_cast<void*>(future))
+        ->registerReleasableCallback(
+            folly::getGlobalCPUExecutor(), 0, unlockFn, mutex);
```

Comment:
we need to create seperate executor.
Share this executor might hit some deadlock

Reply:
I'm working on execution pool in C side, in next PR which will use async cgo in search/query, it will be fixed.